cpp/src/table/row_operators.cu

```diff
@@ -373,51 +385,391 @@ namespace row {
+  // `transformed_input` will produce the same result as sorting `input` regardless of sorting
+  // order (ASC or DESC).
+  auto const compute_ranks = [&](auto const& input) {
+    return cudf::detail::rank(input,
+                              rank_method::DENSE,
+                              order::ASCENDING,
+                              null_policy::EXCLUDE,
+                              column_null_order,
+                              false /*percentage*/,
+                              stream,
+                              default_mr);
+  };
+
+  std::vector<std::unique_ptr<column>> out_cols_lhs;
+  std::vector<std::unique_ptr<column>> out_cols_rhs;
+
+  if (lhs.type().id() == type_id::LIST) {
+    auto const child_lhs = cudf::lists_column_view{lhs}.get_sliced_child(stream);
+
+    // Found a lists-of-structs column.
+    if (child_lhs.type().id() == type_id::STRUCT) {
+      if (rhs_opt) {  // rhs table is available
+        auto const child_rhs = cudf::lists_column_view{rhs_opt.value()}.get_sliced_child(stream);
+        auto const concatenated_children = cudf::detail::concatenate(
+          std::vector<column_view>{child_lhs, child_rhs}, stream, default_mr);
+
+        auto const ranks        = compute_ranks(concatenated_children->view());
+        auto const ranks_slices = cudf::detail::slice(
+          ranks->view(),
+          {0, child_lhs.size(), child_lhs.size(), child_lhs.size() + child_rhs.size()},
+          stream);
+
+        out_cols_lhs.emplace_back(std::make_unique<column>(ranks_slices.front()));
+        out_cols_rhs.emplace_back(std::make_unique<column>(ranks_slices.back()));
+
+        auto transformed_lhs = replace_child(lhs, out_cols_lhs.back()->view(), out_cols_lhs);
+        auto transformed_rhs =
+          replace_child(rhs_opt.value(), out_cols_rhs.back()->view(), out_cols_rhs);
+
+        return {std::move(transformed_lhs),
+                std::optional<column_view>{std::move(transformed_rhs)},
+                std::move(out_cols_lhs),
+                std::move(out_cols_rhs)};
+      } else {  // rhs table is not available
+        out_cols_lhs.emplace_back(compute_ranks(child_lhs));
+        auto transformed_lhs = replace_child(lhs, out_cols_lhs.back()->view(), out_cols_lhs);
+
+        return {std::move(transformed_lhs),
+                std::nullopt,
+                std::move(out_cols_lhs),
+                std::move(out_cols_rhs)};
+      }
+    }
+    // Found a lists-of-lists column.
+    else if (child_lhs.type().id() == type_id::LIST) {
+      auto const child_rhs_opt =
+        rhs_opt
+          ? std::optional<column_view>{cudf::lists_column_view{rhs_opt.value()}.get_sliced_child(
+              stream)}
+          : std::nullopt;
+
+      // Recursively call transformation on the child column.
+      auto [new_child_lhs, new_child_rhs_opt, out_cols_child_lhs, out_cols_child_rhs] =
+        transform_lists_of_structs(child_lhs, child_rhs_opt, column_null_order, stream);
```

Comment:
The original input should be passed in `decompse_struct` before flowing into this function. So basically this function cannot see any structs-of-lists column. It can only see structs-of-structs-of..... of basic types, or lists-of-lists, or lists-of-structs.

Reply:
Ah right and `decompose_structs` will find structs at any level of nesting, even inside lists, right? That's the key here is ensuring that even when there's mixed nesting every level gets processed correctly (or at least, whatever constitutes the "leaf" nodes according to that function).

Reply:
> `decompose_structs` will find structs at any level of nesting, even inside lists, right?

Yes. All possible mixed nesting at any level will be processed to make sure this function will not miss any cases.